service/src/main/kotlin/fi/espoo/evaka/shared/db/Database.kt

```diff
@@ -0,0 +1,190 @@
+            }
+        }
+
+        /**
+         * Starts a transaction, runs the given function, and commits or rolls back the transaction depending on whether
+         * the function threw an exception or not.
+         *
+         * Throws `IllegalStateException` if this database connection is already in read mode or a transaction.
+         */
+        fun <T> transaction(f: (db: Transaction) -> T): T {
+            threadId.assertCurrentThread()
+            check(!handle.isInTransaction) { "Already in a transaction" }
+            return handle.transaction { f(Transaction(it)) }
+        }
+
+        companion object {
+            /**
+             * Wraps an existing raw JDBI handle into a `Connection` object.
+             *
+             * This is mostly intended for tests where the main API can't be used. Use *very* sparingly!
+             */
+            fun wrap(handle: Handle): Connection {
+                check(!handle.isInTransaction) { "Wrapped handle must not have an active transaction" }
+                check(!handle.isReadOnly) { "Wrapped handle must not be read-only" }
+                return Connection(ThreadId(), handle)
+            }
+        }
+
+        override fun close() {
+            threadId.assertCurrentThread()
+            handle.close()
+        }
+    }
+
+    /**
+     * A single database connection in read mode.
+     *
+     * Tied to the thread that created it, and throws `IllegalStateException` if used in the wrong thread.
+     */
+    open class Read internal constructor(val handle: Handle) {
+        fun createQuery(@Language("sql") sql: String): Query = handle.createQuery(sql)
+
+        companion object {
+            /**
+             * Wraps an existing raw JDBI handle into a `Read` object.
+             *
+             * This is mostly intended for tests where the main API can't be used. Use *very* sparingly!
+             */
+            fun wrap(handle: Handle): Read {
+                check(handle.isInTransaction) { "Wrapped handle must have an active transaction" }
+                check(handle.isReadOnly) { "Wrapped handle must be read-only" }
+                return Read(handle)
+            }
+        }
+    }
+
+    /**
+     * A single database connection running a transaction.
+     *
+     * Tied to the thread that created it, and throws `IllegalStateException` if used in the wrong thread.
+     */
+    class Transaction internal constructor(handle: Handle) : Database.Read(handle) {
+        private var savepointId: Long = 0
+
+        fun nextSavepoint(): String = "savepoint-${savepointId++}"
+        fun createUpdate(@Language("sql") sql: String): Update = handle.createUpdate(sql)
+        fun prepareBatch(@Language("sql") sql: String): PreparedBatch = handle.prepareBatch(sql)
+        fun execute(@Language("sql") sql: String, vararg args: Any): Int = handle.execute(sql, *args)
+
+        inline fun <reified T> subTransaction(crossinline f: () -> T): T {
```

Comment:
Can you explain why is this function `inline` by the way? What's the practical reason? I didn't really understand Kotlin's documentation about `crossinline` parameters.

Reply:
Nevermind, I got it now.